scripts/demangle.py

```diff
@@ -1,104 +1,170 @@
 import os
 import json
-import glob
 import re
-import ctypes
+import docker
+from logging import error, info, warning
+import cpp_demangle
+import argparse
+from typing import Optional
+from common import OSSFUZZ_SCRIPTS_HOME
 
-json_path = '../dump/augeas'
-proj_path = '../oss-fuzz/projects/augeas'
-file_names = [f for f in os.listdir(json_path) if f.endswith('.json') and os.path.isfile(json_path+'/'+f)]
-print('Looking for json file: ',file_names)
-def extract_func_code(func_name, file_path):
-    with open(file_path, 'r') as file:
-        code_content = file.read()
-    # init regular expression
-    match_func_init = re.search(".*"+func_name,code_content)
+
+# Extract specific function
+def extract_func_code(
+    func_name: str, code_content: str, if_c_code: bool
+) -> Optional[str]:
+    ori_func_name = func_name
+    # Transliteration special characterss, replace special characters with \ ahead of themselves
+    pattern = r"[\[\].,{}()\W_]"
+    func_name = re.sub(pattern, r"\\\g<0>", func_name)
+    # If C code, there is no demangle process, only have function name
+    # Thus manually add parphsis to match the function
+    if if_c_code:
+        # Searching string that start with anything followed with function name+()+{
+        match_func_init = re.search(
+            f"(?:.*\n|.*){func_name}\s*\([^;]*\)(?:\s*|\n)\\{{\n", code_content
+        )
+    else:
+        match_func_init = re.search(f".*{func_name}[^;]*\n", code_content)
+    # If found matched function
     if match_func_init:
-        func_init, func_start = match_func_init.group(), match_func_init.end()
+        # The regular expression end with {\n, thus I need to move ahead three char to get the full function namespace.
+        func_init, func_start = match_func_init.group(), match_func_init.end() - 3
     else:
-        #raise ValueError(f"No match found")
-        print(f"ERROR: No match function '{func_name}' found")
-        return f"ERROR: No match function '{func_name}' found"
-    
+        warning(f" No match function {ori_func_name} found")
+        return None
+    # Match braces to get function code
     func_now, open_braces, flag = func_start, 0, 1
-
     while (open_braces > 0 or flag) and func_now < len(code_content):
```

Comment:
This code may be incorrect if `code_content` doesn't have braces. You can solve this (and remove unncessary flag) by setting `func_now` to the pos of the first left braces.

```py
func_now, open_braces = code_content.find("{"), 1
if (func_now == -1):
   err()
while func_now < len(code_content:
   ...
```

Reply:
If `code_content` doesn't have braces, the open_braces is always 0. When it goes over all `code_content` it will end.